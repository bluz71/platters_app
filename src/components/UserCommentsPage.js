import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import axios from 'axios';
import pluralize from 'pluralize';
import numeral from 'numeral';
import { Row, Col, PageHeader } from 'react-bootstrap';
import FontAwesome from 'react-fontawesome';
import { API_HOST } from '../config';
import pageProgress from '../helpers/pageProgress';
import infiniteScroll from '../helpers/infiniteScroll';
import toastAlert from '../helpers/toastAlert';
import CommentsList from './CommentsList';
import '../styles/UserCommentsPage.css';

class UserCommentsPage extends Component {
  constructor(props) {
    super(props);

    this.userSlug         = props.match.params.id;
    this.commentsEndPoint = `${API_HOST}/comments/${this.userSlug}.json`;
    this.loaded           = false;
    this.waiting          = false;
    this.pageProgress     = new pageProgress();

    this.state = {
      comments: new Map(),
      pagination: {},
      notFound: false,
      error: null
    };

    // Bind 'this' for callback functions.
    this.handleScroll  = this.handleScroll.bind(this);
    this.handlePageEnd = this.handlePageEnd.bind(this);
  }

  componentDidMount() {
    window.onscroll = this.handleScroll;
    this.getComments(this.commentsEndPoint);
  }

  componentWillUnmount() {
    window.onscroll = null;
  }

  handleScroll() {
    infiniteScroll(this.state.pagination, this.handlePageEnd);
  }

  handlePageEnd() {
    // Page has been scrolled to the end, hence retrieve the next set of
    // comments.

    // Disable scroll handling whilst records are being retrieved.
    window.onscroll = null;
    const commentsPageEndPoint
      = `${this.commentsEndPoint}?page=${this.state.pagination.next_page}`;
    this.waiting = true;
    this.forceUpdate(); // Render spinner
    this.getComments(commentsPageEndPoint);
  }

  getComments(commentsEndPoint) {
    axios.get(commentsEndPoint)
      .then(response => {
        this.loaded = this.pageProgress.done();
        this.waiting = false;
        if (!window.onscroll) {
          // Re-enable scroll handling now that the records have been
          // retrieved.
          window.onscroll = this.handleScroll;
        }
        // console.log(`UserCommentsPage data: ${JSON.stringify(response, null, 2)}`)
        this.setState({
          comments: new Map([
            ...this.state.comments,
            ...response.data.comments.map(c => [c.id, c])
          ]),
          pagination: response.data.pagination,
        });
      }).catch(error => {
        this.pageProgress.done();
        if (error.response && error.response.status === 404) {
          this.setState({ notFound: true });
        }
        else {
          this.setState({ error: error });
        }
      });
  }

  commentsRetrieved() {
    if (!this.loaded || this.state.notFound || this.state.error) {
      return false;
    }
    else {
      return true;
    }
  }

  renderHeader() {
    if (this.state.notFound) {
      toastAlert(`User ${this.userSlug} does not exist`);
      return <Redirect to="/" />;
    }
    if (this.state.error) {
      toastAlert('Connection failure, please retry again soon');
      return <Redirect to="/" />;
    }

    const count = this.state.pagination.total_count;
    const commentsCount = numeral(count).format('0,0');

    return (
      <PageHeader>
        Comments {this.commentsRetrieved()
            && <small>({commentsCount} {pluralize('Comment', count)})</small>}
      </PageHeader>
    );
  }

  renderComments() {
    return (
      <CommentsList comments={this.state.comments} />
    );
  }

  renderSpinner() {
    if (this.waiting) {
      return (
        <div className="WaitingSpinner">
          <FontAwesome name="spinner" spin pulse />
        </div>
      );
    }
  }

  render() {
    this.pageProgress.start();

    return (
      <Row>
        <Col md={10} mdOffset={1}>
          <div className="UserComments">
            {this.renderHeader()}
            {this.renderComments()}
            {this.renderSpinner()}
          </div>
        </Col>
      </Row>
    );
  }
}

export default UserCommentsPage;
