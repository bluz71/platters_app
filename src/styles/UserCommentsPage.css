.UserComments .page-header h1 {
  font-size: 26px;
}

.UserComments .page-header h1 small {
  font-size: 65%;
}

.UserComments .LoadingSpinner {
  text-align: center;
  font-size: 2em;
  color: #777;
  margin-bottom: 0.5em;
}
