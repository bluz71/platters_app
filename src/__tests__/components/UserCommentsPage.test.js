import React from 'react';
import { shallow } from 'enzyme';
import UserCommentsPage from '../../components/UserCommentsPage';


// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856
const flushPromises = () => new Promise(resolve => setImmediate(resolve));

describe('<UserCommentsPage />', () => {
  it('renders user comments page', async () => {
    const match = { params: { id: 'joe' } };
    const wrapper = shallow(<UserCommentsPage  match={match}/>);
    await flushPromises();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
