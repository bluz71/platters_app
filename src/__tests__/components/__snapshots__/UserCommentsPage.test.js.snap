// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserCommentsPage /> handles invalid user 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    md={10}
    mdOffset={1}
  >
    <div
      className="UserComments"
    >
      <Redirect
        push={false}
        to="/"
      />
      <CommentsList
        comments={Map {}}
        onDeleteComment={[Function]}
      />
    </div>
  </Col>
</Row>
`;

exports[`<UserCommentsPage /> renders all user comments when scrolled 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    md={10}
    mdOffset={1}
  >
    <div
      className="UserComments"
    >
      <PageHeader
        bsClass="page-header"
      >
        Comments 
        <small>
          (
          3
           
          Comments
          )
        </small>
      </PageHeader>
      <CommentsList
        comments={
          Map {
            1 => Object {
              "artist": "ABC",
              "body": "First comment",
              "created_at": "<time datetime=\\"2018-05-09T04:05:00Z\\" data-local=\\"time-ago\\">May  9, 2018  4:05am</time>",
              "for": "Album",
              "gravatar_url": "https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon",
              "id": 1,
              "name": "ABC",
              "path": "artist/abc/album/abc",
              "user_name": "joe",
              "user_slug": "joe",
            },
            2 => Object {
              "artist": "DEF",
              "body": "Second comment",
              "created_at": "<time datetime=\\"2018-05-05T05:34:56Z\\" data-local=\\"time-ago\\">May  5, 2018  5:34am</time>",
              "for": "Album",
              "gravatar_url": "https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon",
              "id": 2,
              "name": "DEF",
              "path": "artist/def/album/def",
              "user_name": "joe",
              "user_slug": "joe",
            },
            3 => Object {
              "artist": "XYZ",
              "body": "Third comment",
              "created_at": "<time datetime=\\"2018-05-02T08:15:00Z\\" data-local=\\"time-ago\\">May  2, 2018  8:15am</time>",
              "for": "Album",
              "gravatar_url": "https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon",
              "id": 3,
              "name": "XYZ",
              "path": "artist/xyz/album/xyz",
              "user_name": "joe",
              "user_slug": "joe",
            },
          }
        }
        onDeleteComment={[Function]}
      />
    </div>
  </Col>
</Row>
`;

exports[`<UserCommentsPage /> renders user comments page 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    md={10}
    mdOffset={1}
  >
    <div
      className="UserComments"
    >
      <PageHeader
        bsClass="page-header"
      >
        Comments 
        <small>
          (
          3
           
          Comments
          )
        </small>
      </PageHeader>
      <CommentsList
        comments={
          Map {
            1 => Object {
              "artist": "ABC",
              "body": "First comment",
              "created_at": "<time datetime=\\"2018-05-09T04:05:00Z\\" data-local=\\"time-ago\\">May  9, 2018  4:05am</time>",
              "for": "Album",
              "gravatar_url": "https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon",
              "id": 1,
              "name": "ABC",
              "path": "artist/abc/album/abc",
              "user_name": "joe",
              "user_slug": "joe",
            },
            2 => Object {
              "artist": "DEF",
              "body": "Second comment",
              "created_at": "<time datetime=\\"2018-05-05T05:34:56Z\\" data-local=\\"time-ago\\">May  5, 2018  5:34am</time>",
              "for": "Album",
              "gravatar_url": "https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon",
              "id": 2,
              "name": "DEF",
              "path": "artist/def/album/def",
              "user_name": "joe",
              "user_slug": "joe",
            },
          }
        }
        onDeleteComment={[Function]}
      />
    </div>
  </Col>
</Row>
`;
